Remove usage of IDocumentStorageService.write

kian-thompson opened the pull request:
Also removes all implementations of the `write` method from classes who implement `IDocumentStorageService`
#9250

kian-thompson committed:
Remove usage of IDocumentStorageService.write
- ** Note: this does not remove implementations of the method from classes who implement IDocumentStorageService

vladsud commented:
Looks reasonable (as a started). Looking forward to complete client changes.

kian-thompson commented:
@vladsud Should I create a separate PR for removing the `write` method? Then once that PR is merged I can consume the new `@fluidframework/driver-definitions` version in this PR.

vladsud commented:
Yes, but first I think we should prove to ourselves that we can indeed remove it. I.e. there is no usage of this API (other than various proxy implementations). So I think developing this draft PR to full completion and proving to yourself that it's complete (i.e. it compiles just fine without write method on interface) is a must before actual removal.
It might be hard to find all implementations (I think we can find most of them, but if we miss one - it's not a big deal). But finding all usage (which is hopefully zero) is a must.

kian-thompson committed:
Remove implementations of write method

kian-thompson committed:
Updated md files from build

kian-thompson commented:
@vladsud I've gone ahead and removed all implementations of the `write` method in this PR. I adjusted the compiled .js files locally and the only build failures are for backwards compatibility on various interfaces and classes. Commenting out all these backwards compat tests sees the build succeed.

vladsud commented:
LGTM! Please proceed with definition change (removal of method from interface).
I'm surprised you are hitting compat test issues, but we can glance one definition changes are in.

kian-thompson commented:
> LGTM! Please proceed with definition change (removal of method from interface). I'm surprised you are hitting compat test issues, but we can glance one definition changes are in.

I assume I need to create a separate PR to remove the method? Then update the version for `@fluidframework/driver-definitions` in this PR once that is built and pushed?

kian-thompson committed:
Update @fluidframework/driver-definitions dependencies

kian-thompson committed:
Revert "Update @fluidframework/driver-definitions dependencies"

This reverts commit fc47b77249c8ea8fdc2c2d5a11a812e132500809.

kian-thompson committed:
Merge branch 'next' into 9250-remove-idocumentstorageservicewrite

kian-thompson committed:
Bump dependencies version

      @fluidframework/driver-definitions -> ^0.47.1000-0

kian-thompson committed:
Merge pull request #1 from kian-thompson/dep_1650664621062

Bump dependencies version

kian-thompson committed:
Fix back compat issue

vladsud commented on common/lib/container-definitions/package.json:
```diff
@@ -74,6 +74,9 @@
     "broken": {
       "InterfaceDeclaration_IRuntime": {
         "forwardCompat": false
+      },
```
Curious, why do you need to change it? Seems like interface is not changed

kian-thompson replied:
The `storage` property on `IContainerContext` is `IDocumentStorageService`, so the back compat test fails since the `IDocumentStorageService` signatures aren't the same

vladsud replied:
Hm, this feels weird as I'd not expect version check to be nested like that. If it goes recursive, a change anywhere will cause a lot of types to be different. @anthony-murphy, any chance you can glance?

kian-thompson replied:
@vladsud Every time I build it finds new things broken with back compat. Because it seems to be recursive, it will be a lot bigger of a change than I anticipated. I'm currently very tempted to revert my commit that removed this method.

vladsud replied:
This feels wrong. Can you please ping Tony? I do not think it should be recursive.
Maybe other way to say it - we need to be able to make changes like that. Maybe here we can punt, but it would be nuts if we can't change interfaces rather easily.

anthony-murphy replied:
this is the way it currently works, and represents the real break you've made. The type tests don't do anything fancy, they just check if the old type can be used in place of the new type and vis versa via the typescript compiler.

anthony-murphy replied:
to me this points to issues is our code. why do we expose the same thing in so many places?

kian-thompson replied:
@anthony-murphy Take `IProvideContainerRuntime` for example:
1. `IProvideContainerRuntime` has a property with type `IContainerRuntime`
2. `IContainerRuntime` has a property with type `IDocumentStorageService`
3. And thus, back compat for `IProvideContainerRuntime` breaks

Changing low-level interfaces seems to spread quite a bit even though no one is really consuming this `write` method. There are a number of places that implement the `IDocumentStorageService`, so the breaking spreads very quickly (which could be expected).

vladsud replied:
I'm sure it's not the last time we will hit issues like that. It feels like one of the two should be possible:
1. A way for developers to ignore all errors (without updating 100 places)
2. A script that developer can run to auto-generate package.json changes, similar to api.md generation process.

I personally would favor # 2, and I find I need it even for trivial compat changes. It feels like it's in same bucket as linting/formatting changes - it's so much easier to work in Bohemia repo where changes are done automatically, and dev responsibility is to review them.

kian-thompson committed:
Fix more back compat

kian-thompson committed:
Fix more back compat

kian-thompson committed:
You guessed it. More back compat

kian-thompson committed:
Merge branch 'next' into 9250-remove-idocumentstorageservicewrite

kian-thompson committed:
Bump dependencies version

   @fluidframework/container-definitions -> ^0.49.1000-0

kian-thompson committed:
Merge pull request #2 from kian-thompson/dep_1650913683059

Bump dependencies version

kian-thompson committed:
Remove implementation of "write" method for SerializedSnapshotStorage

kian-thompson committed:
Use specific pre-release for container-definitions

kian-thompson committed:
Fix more back compat

kian-thompson committed:
Hopefully the last back compat fix
